File: ros2/rs_pubsub/ros2_benchmark_rs/src/listener.rs
//! ROS 2 Rust listener using rclrs — mirrors dora-rs/rs-latency/sink.
//!
//! Subscribes to `/topic`, measures latency per size bracket, and
//! writes results to `time.csv` with the same schema as
//! `dora-rs/rs-latency/timer.csv`.

use anyhow::{Error, Result};
use csv::Writer;
use rclrs::*;
use std::time::Duration;

const LANGUAGE: &str = "Rust";
const NAME_DEFAULT: &str = "ROS 2 Rust (rclrs)";
const PLATFORM: &str = "COMPUTER_PERF";

fn clock_monotonic_ns() -> u64 {
    let mut ts = libc::timespec {
        tv_sec: 0,
        tv_nsec: 0,
    };
    // SAFETY: clock_gettime with CLOCK_MONOTONIC is always safe on Linux.
    unsafe {
        libc::clock_gettime(libc::CLOCK_MONOTONIC, &mut ts);
    }
    (ts.tv_sec as u64) * 1_000_000_000 + (ts.tv_nsec as u64)
}

struct State {
    current_size: usize,
    latencies: Vec<Duration>,
    date: String,
}

fn main() -> Result<(), Error> {
    let context = Context::default_from_env()?;
    let mut executor = context.create_basic_executor();
    let node = executor.create_node("ros2_benchmark_listener")?;

    let worker = node.create_worker::<State>(State {
        current_size: 0,
        latencies: Vec::new(),
        date: epoch_string(),
    });

    let _subscription = worker.create_subscription::<std_msgs::msg::UInt64MultiArray, _>(
        "topic",
        move |state: &mut State, msg: std_msgs::msg::UInt64MultiArray| {
            let t_received = clock_monotonic_ns();
            let size_bytes = msg.data.len() * 8;

            if size_bytes != state.current_size {
                if !state.latencies.is_empty() {
                    record_results(state.current_size, &mut state.latencies, &state.date);
                }
                state.current_size = size_bytes;
                state.latencies.clear();
            }

            let t_send = msg.data[0];
            let latency = Duration::from_nanos(t_received.saturating_sub(t_send));
            state.latencies.push(latency);
        },
    )?;

    println!("Waiting for messages...");
    executor.spin(SpinOptions::default()).first_error()?;
    Ok(())
}

fn epoch_string() -> String {
    use std::time::SystemTime;
    let secs = SystemTime::now()
        .duration_since(SystemTime::UNIX_EPOCH)
        .map(|d| d.as_secs())
        .unwrap_or(0);
    format!("epoch+{secs}")
}

fn percentile(sorted: &[Duration], pct: f64) -> Duration {
    if sorted.is_empty() {
        return Duration::ZERO;
    }
    let idx = ((pct / 100.0) * (sorted.len() - 1) as f64).round() as usize;
    sorted[idx.min(sorted.len() - 1)]
}

fn record_results(size_bytes: usize, latencies: &mut Vec<Duration>, date: &str) {
    latencies.sort();
    let n = latencies.len() as u32;
    let avg = latencies.iter().sum::<Duration>() / n;
    let p50 = percentile(latencies, 50.0);
    let p90 = percentile(latencies, 90.0);
    let p99 = percentile(latencies, 99.0);

    let path = std::env::var("CSV_TIME_FILE").unwrap_or_else(|_| "time.csv".to_string());
    let file = std::fs::OpenOptions::new()
        .create(true)
        .append(true)
        .open(path)
        .expect("open csv");
    let mut wtr = Writer::from_writer(file);

    let name = std::env::var("NAME").unwrap_or_else(|_| NAME_DEFAULT.to_string());
    wtr.write_record([
        date.to_string(),
        LANGUAGE.to_string(),
        PLATFORM.to_string(),
        name,
        size_bytes.to_string(),
        avg.as_micros().to_string(),
        p50.as_micros().to_string(),
        p90.as_micros().to_string(),
        p99.as_micros().to_string(),
        n.to_string(),
    ])
    .expect("write csv row");

    println!(
        "size={size_bytes}  avg={}µs  p50={}µs  p90={}µs  p99={}µs  n={n}",
        avg.as_micros(),
        p50.as_micros(),
        p90.as_micros(),
        p99.as_micros()
    );
}

File: ros2/cpp_pubsub/src/listener.cpp
// ROS 2 C++ listener — mirrors dora-rs/rs-latency/sink.
//
// Subscribes to /topic, measures latency per size bracket, and writes
// results to time.csv with the same schema as dora-rs/rs-latency/timer.csv.

#include <algorithm>
#include <chrono>
#include <cmath>
#include <cstdint>
#include <cstdlib>
#include <fstream>
#include <iostream>
#include <string>
#include <vector>

#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/u_int64_multi_array.hpp"

static uint64_t clock_monotonic_ns() {
  struct timespec ts;
  clock_gettime(CLOCK_MONOTONIC, &ts);
  return static_cast<uint64_t>(ts.tv_sec) * 1'000'000'000ULL +
         static_cast<uint64_t>(ts.tv_nsec);
}

static uint64_t percentile_us(std::vector<uint64_t> &sorted, double pct) {
  if (sorted.empty())
    return 0;
  size_t idx = static_cast<size_t>(
      std::round((pct / 100.0) * static_cast<double>(sorted.size() - 1)));
  if (idx >= sorted.size())
    idx = sorted.size() - 1;
  return sorted[idx];
}

static void record_results(const std::string &date, size_t size_bytes,
                           std::vector<uint64_t> &latencies_us) {
  std::sort(latencies_us.begin(), latencies_us.end());
  uint64_t n = latencies_us.size();
  uint64_t sum = 0;
  for (auto v : latencies_us)
    sum += v;
  uint64_t avg = sum / n;
  uint64_t p50 = percentile_us(latencies_us, 50.0);
  uint64_t p90 = percentile_us(latencies_us, 90.0);
  uint64_t p99 = percentile_us(latencies_us, 99.0);

  const char *csv_env = std::getenv("CSV_TIME_FILE");
  std::string csv_path = csv_env ? csv_env : "time.csv";

  const char *name_env = std::getenv("NAME");
  std::string name = name_env ? name_env : "ROS 2 C++";

  std::ofstream f(csv_path, std::ios::app);
  f << date << ",C++,COMPUTER_PERF," << name << "," << size_bytes << "," << avg
    << "," << p50 << "," << p90 << "," << p99 << "," << n << "\n";

  std::cout << "size=" << size_bytes << "  avg=" << avg << "us  p50=" << p50
            << "us  p90=" << p90 << "us  p99=" << p99 << "us  n=" << n
            << std::endl;
}

class Listener : public rclcpp::Node {
public:
  Listener()
      : Node("cpp_benchmark_listener"), current_size_(0) {
    // Get a rough ISO timestamp for CSV labelling.
    auto now = std::chrono::system_clock::now();
    auto t = std::chrono::system_clock::to_time_t(now);
    char buf[64];
    std::strftime(buf, sizeof(buf), "%Y-%m-%dT%H:%M:%S", std::localtime(&t));
    date_ = buf;

    subscription_ = this->create_subscription<std_msgs::msg::UInt64MultiArray>(
        "topic", 10,
        [this](const std_msgs::msg::UInt64MultiArray::SharedPtr msg) {
          this->callback(msg);
        });
  }

  ~Listener() {
    if (!latencies_us_.empty()) {
      record_results(date_, current_size_, latencies_us_);
    }
  }

private:
  void callback(const std_msgs::msg::UInt64MultiArray::SharedPtr msg) {
    uint64_t t_received = clock_monotonic_ns();
    size_t size_bytes = msg->data.size() * 8;

    if (size_bytes != current_size_) {
      if (!latencies_us_.empty()) {
        record_results(date_, current_size_, latencies_us_);
      }
      current_size_ = size_bytes;
      latencies_us_.clear();
    }

    uint64_t t_send = msg->data[0];
    uint64_t latency_ns = t_received - t_send;
    latencies_us_.push_back(latency_ns / 1000); // ns -> us
  }

  rclcpp::Subscription<std_msgs::msg::UInt64MultiArray>::SharedPtr
      subscription_;
  size_t current_size_;
  std::vector<uint64_t> latencies_us_;
  std::string date_;
};

int main(int argc, char *argv[]) {
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<Listener>());
  rclcpp::shutdown();
  return 0;
}
